onmt/bin/build_vocab.py

```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python
+"""Get vocabulary coutings from transformed corpora samples."""
+from onmt.utils.logging import init_logger
+from onmt.utils.misc import set_random_seed
+from onmt.dynamic.parse import DynamicArgumentParser
+from onmt.dynamic.opts import dynamic_prepare_opts
+from onmt.dynamic.corpus import save_transformed_sample
+from onmt.dynamic.transforms import make_transforms, get_transforms_cls
+
+
+def build_vocab_main(opts):
+    """Apply transforms to samples of specified data and build vocab from it.
+
+    Transforms that need vocab will be disabled in this.
+    Built vocab is saved in plain text format as following and can be pass as
+    `-src_vocab` (and `-tgt_vocab`) when training:
+    ```
+    <tok_0>\t<count_0>
+    <tok_1>\t<count_1>
+    ```
+    """
+
+    DynamicArgumentParser.validate_prepare_opts(opts)
+    assert opts.n_sample > 1, f"Illegal argument n_sample={opts.n_sample}."
+
+    logger = init_logger()
+    set_random_seed(opts.seed, False)
+    transforms_cls = get_transforms_cls(opts._all_transform)
+    fields = None
+
+    transforms = make_transforms(opts, transforms_cls, fields)
+
+    logger.info(f"Counter vocab from {opts.n_sample} samples.")
+    src_counter, tgt_counter = save_transformed_sample(
+        opts, transforms, n_sample=opts.n_sample, build_vocab=True)
+
+    logger.info(f"Counters src:{len(src_counter)}")
+    logger.info(f"Counters tgt:{len(tgt_counter)}")
+    if opts.share_vocab:
+        src_counter += tgt_counter
+        tgt_counter = src_counter
+        logger.info(f"Counters after share:{len(src_counter)}")
+
+    def save_counter(counter, save_path):
+        with open(save_path, "w") as fo:
+            for tok, count in counter.most_common():
+                fo.write(tok + "\t" + str(count) + "\n")
```

Comment:
We need to add a condition somewhere for `tok != ""`.
Not sure why but in some instances we get some empty string as a token:
```
...
aberration	114
	114
infirm	114
...
```
which causes this error when reading it when training:
```
Traceback (most recent call last):
  File "onmt/bin/train.py", line 118, in <module>
    main()
  File "onmt/bin/train.py", line 114, in main
    train(opt)
  File "onmt/bin/train.py", line 53, in train
    prepare_fields_transforms(opt)
  File "onmt/bin/train.py", line 32, in prepare_fields_transforms
    opt, src_specials=specials['src'], tgt_specials=specials['tgt'])
  File "/nas-labs/MT/pytorchwork/Recipes/toy-fh/OpenNMT-py/onmt/dynamic/fields.py", line 34, in build_dynamic_fields
    min_freq=opts.src_words_min_frequency)
  File "/nas-labs/MT/pytorchwork/Recipes/toy-fh/OpenNMT-py/onmt/inputters/inputter.py", line 281, in _load_vocab
    for token, count in vocab:
ValueError: not enough values to unpack (expected 2, got 1)
```

Reply:
> We need to add a condition somewhere for `tok != ""`.
> Not sure why but in some instances we get some empty string as a token:
> 
> ```
> ...
> aberration	114
> 	114
> infirm	114
> ...

Did this happen in `build_vocab.py`?

Reply:
Yes. Probably some `.split()` that produces some empty strings at some point.
Ideally we could track down where it stems from in `build_vocab`, I'll have a look.
But it wouldn't hurt to handle the case as well in the "vocab loading" part.

Reply:
> Yes. Probably some .split() that produces some empty strings at some point.

This is actually caused by some empty lines in the data.
In `onmt.dynamic.corpus.save_transformed_sample`, counters are updated with `.split(" ")`, and it seems that:
```python
"".split() # gives []
"".split(" ") # gives [""]
```

Reply:
Addressed in 6da1fa7 of #21.